_posts/2024-10-03-pixi-basics.md

```diff
@@ -0,0 +1,358 @@
+  Range (min … max):   24.182 s … 25.921 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv samtools && rm -rf condaenv
+
+  Time (mean ± σ):     35.740 s ±  0.770 s    [User: 6.519 s, System: 5.505 s]
+  Range (min … max):   34.609 s … 36.290 s    4 runs
+ 
+Summary
+  'pixi add samtools && pixi install && pixi remove samtools' ran
+    1.45 ± 0.06 times faster than 'conda create -y -p condaenv samtools && rm -rf condaenv'
+
+Benchmark 1: pixi add vadr && pixi install && pixi remove vadr
+
+  Time (mean ± σ):     70.947 s ±  0.554 s    [User: 23.427 s, System: 37.370 s]
+  Range (min … max):   70.519 s … 71.733 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv vadr && rm -rf condaenv
+
+  Time (mean ± σ):     127.758 s ±  7.067 s    [User: 18.144 s, System: 16.139 s]
+  Range (min … max):   122.523 s … 137.886 s    4 runs
+ 
+Summary
+  'pixi add vadr && pixi install && pixi remove vadr' ran
+    1.80 ± 0.10 times faster than 'conda create -y -p condaenv vadr && rm -rf condaenv'
+
+Benchmark 1: pixi add ska && pixi install && pixi remove ska
+
+  Time (mean ± σ):      5.076 s ±  0.149 s    [User: 5.854 s, System: 1.689 s]
+  Range (min … max):    4.948 s …  5.283 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv ska && rm -rf condaenv
+
+  Time (mean ± σ):      8.831 s ±  0.466 s    [User: 4.585 s, System: 2.413 s]
+  Range (min … max):    8.215 s …  9.262 s    4 runs
+ 
+Summary
+  'pixi add ska && pixi install && pixi remove ska' ran
+    1.74 ± 0.10 times faster than 'conda create -y -p condaenv ska && rm -rf condaenv'
+
+Benchmark 1: pixi add pangolin && pixi install && pixi remove pangolin
+
+  Time (mean ± σ):     160.734 s ±  5.329 s    [User: 41.234 s, System: 105.102 s]
+  Range (min … max):   154.892 s … 167.771 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv pangolin && rm -rf condaenv
+
+  Time (mean ± σ):     321.362 s ±  5.470 s    [User: 31.665 s, System: 39.065 s]
+  Range (min … max):   315.114 s … 328.384 s    4 runs
+ 
+Summary
+  'pixi add pangolin && pixi install && pixi remove pangolin' ran
+    2.00 ± 0.07 times faster than 'conda create -y -p condaenv pangolin && rm -rf condaenv'
+Benchmark 1: pixi add ivar && pixi install && pixi remove ivar
+
+  Time (mean ± σ):     24.782 s ±  0.354 s    [User: 7.629 s, System: 7.915 s]
+  Range (min … max):   24.505 s … 25.275 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv ivar && rm -rf condaenv
+
+  Time (mean ± σ):     37.005 s ±  0.452 s    [User: 6.882 s, System: 5.673 s]
+  Range (min … max):   36.575 s … 37.606 s    4 runs
+ 
+Summary
+  'pixi add ivar && pixi install && pixi remove ivar' ran
+    1.49 ± 0.03 times faster than 'conda create -y -p condaenv ivar && rm -rf condaenv'
+
+```
+
+</details>
+
+For the complicated installation, pixi is 2.21x faster than conda.
+For individual installations, it is consistently more than 1.4x faster than conda.
+
+## Day to day operations
+
+Why would you even go through all this?
```

Comment:
Perhaps could be stressing even more how the localized storage of the environment aligns particularly well with reproducibility and observability principles, as it encourages to keep things close and linked together similar to how keeping documentation in the code folder allows it to follow the life cycle of the code (so that you could in theory update the docs everytime you add a new feature, and even have it in the same commit or PR).
  - For me the biggest benefit is I don't forget what conda environment I used for a specific project, which might cause me to accidentally start a new one, or switch between different ones. Here, the setup is pretty fail-safe.
  - There might be further benefits ... Perhaps it even allows to zip the whole folder and run elsewhere?
  - I personally like especially how it allows you to combine ability of installing a custom python version (as is possible in conda IIUC), with the local nature of poetry environments (which I have otherwise preferred, but which relies on pyenv or something else to install python, and can't install non-python packages).

Reply:
I added this to partially address this comment. I'm not clear on `custom python version` - it sounds like it's more nuanced than what conda would give.  Can you clarify?

> However, with pixi, the environment creation is fast enough that you don't have to save it forever in your project folder.
The initialization is folder/project based too -- it is simple enough to write into your notebook to start with `pixi shell` and to have the `pixi.toml` file as part of your project.
This workflow gets much better if you are in a collaborative environment with multiple hands on the same folder.
Activating the correct environment is pretty much fail-safe because the folder itself dictates the environment.
Or, dare to imagine, packaging up an entire project directory with its dependency software in a single toml file.

Reply:
Looks good, but, I wonder, "packing up an entire project directory [...] in a single toml file", was that meant to be "archive file"? Because defining everything in a single toml file is what you get by default anyway, no?

Reply:
I guess I meant all the data and bespoke scripts, etc, of a project directory _plus_ all the software from conda/pixi. All that software would be represented by the toml file. I have revised it to

> Or, dare to imagine, you could archive an entire project directory with its dependency software in a single toml file.

Is that good?

Reply:
I guess that is good. I just didn't know you could define the data deps via the toml file too. But if you are thinking of a way to do that, then this sounds fine!

Reply:
Ok, I think I just wrote a confusing sentence. Is this better?

> Or, dare to imagine, if/when you archive a bioinformatics project, all the software that you used could be archived into a single toml file.

Reply:
Oh! No, I just mean that the toml file can be used as a stand in to show software and versions that can be installed later with pixi. Equivalent to a `conda freeze`.

Reply:
Aha, I see! Then perhaps a slight adjustment of the wording to something like this might be more accurate?:

> Or, dare to imagine, if/when you archive a bioinformatics project, all the software that you used could be freezed and specified in a single toml file.

Sorry for the confusion! I realize I was imagining something else happening here, causing me to misunderstand.

Reply:
That makes sense - thanks for helping me get to a clear statement on that. I have updated to your suggestion.